### ISSUE TITLE
Previous attempts

### ISSUE THREAD

**Issue by bilderbuchi** [pymeasure/experiment_control_protocol#5]
Our community members have implemented separate approaches to solve the orchestration/coordination problem:

**Comment by bilderbuchi** [comment]
@bklebel:
I think this is in general a very good approach.  In fact, when I put together a system to control a cryostat for low-temperature measurements (mostly resistivity, but under unconventional environmental conditions besides low temperatures), we gradually moved from a monolithic implementation to such a distributed one. If you are interested in this approach, you can also check it out [here](https://github.com/Cryostat-GUI/Cryostat-GUI). 

_short description of our approach:_ 

> In this framework of a program, we use separate processes for communication with individual devices (named `ControlClient`), which then all publish their data via zmq. We have separate GUIs for all of them, however, to measure with the least interruptions, they are actually packed into services which will be restarted once they crash for whatever reason. 
> The second communication channel is through zmq publishers to which all such instrument control clients are subscribed, to receive commands (like to change the temperature). 
> The third channel is a custom request-reply pattern, so that individual states can be requested, and reacted to. 
> 
> Besides multiple instrument control clients, we have a datalogger which is subscribed to the control clients data output, which is a) saved to a logfile, and b) on which we do a few calculations (mainly arithmetic means, slopes, slopes of means etc for temperature stabilisation purposes, the implementation is not very performant or well done, but it serves our purpose), which can in turn be requested through the req-repl pattern. 
>
> In this way, we can have multiple instances sending commands, or reading data, or both. concurrently with all this, all the collected data is pushed to a prometheus+grafana setup, for a simple way to check the instrument's status. All of it is done by a poor PhD student and some fellow BSc students, therefore the quality of both individual pieces and the overall architecture might not be the very best. 

In short, I agree with introducing such a thing into pymeasure, even though I guess I will not be able to convince people in my group to switch away from our current system, nor will I be easily able to convince anybody to use the system we built, and even less extend it. I think, for some types of measurements, it is a more appropriate system than one-script procedures, but for others it might be a bit of an overkill. Introducing just the `datalogger` might be just fine, I think

**Comment by bilderbuchi** [comment]
@bmoneke:

Thank you for sharing your solution. I'm amazed how similar our approaches are (here is mine on [RWTH gitlab](https://git.rwth-aachen.de/nloqo/eg-labor/-/tree/master/DataLogger)).
Nice idea to distribute instrument commands via publish-subscribe patterns, do you distinguish the "receiver" of a command by topic selection?
I use a request-reply pattern to control the datalogger or a couple of individual programs. For a few, it is ok, but for more, you have to assign individual port numbers which becomes a hassle.

In fact, I created a xpub-xsub `proxy_server`, which contains two port numbers (one for subscribing, one for publishing). All individual devices publish their data to this proxy and all data consumers (datalogger instances, a beamprofiler etc.) subscribe to that proxy. Therefore any program can receive the data of any data producer.
An additional proxy server can connect the `proxy_servers` on two computers, such that you can get the data of one computer on another one without telling either the data producers or data consumers that (part) of their correspondents is on another computer.

Maybe it is possible to unify our underlying communication scheme and publish it here on pymeasure, such that improvements work for all participating groups.

EDIT: It is great to have your input, in order to combine the benefits of two approaches and two parts of experience into better software.

**Comment by bilderbuchi** [comment]
@bilderbuchi 
I also made a program in the past (sadly not public) where I had a GUI (autopopulated from instrument properties; with plotting), polling & commanding scheme, logging, data storage, something like procedures (with a smaller command DSL based on [this](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwiS9saR7rL7AhUFQ_EDHbFjCN0QFnoECAgQAQ&url=https%3A%2F%2Fevents.libre.space%2Fevent%2F3%2Fcontributions%2F49%2Fattachments%2F45%2F66%2Flibrecube_pluto.pdf&usg=AOvVaw0gTe-uLmdSS2JH6jDT-yw2)), and instrument layer using pymeasure + others.
Most of these parts communicated with each other via zmq, running in separate threads or processes, and it worked beautifully.

**Comment by bilderbuchi** [comment]
@untzag writes:
I hate to "self-promote", but we've been thinking along these same lines when designing [yaq](https://yaq.fyi/). We even have autogenerated GUIs etc... You might develop some inspiration from what we've done. Happy to share learnings from our journey. My main focus now is integration with Bluesky.

[pyacq](https://github.com/pyacq/pyacq) also sounds similar to what you're thinking of. Might gather inspiration. I think ACQ4 is built on top of pyacq, but I'm not sure. If memory serves PyMoDAQ also has "remote" interfaces that were bolted on after the main program was written.

**Comment by bmoneke** [comment]
In this [gist](https://gist.github.com/bmoneke/fc7be8b848d3afa1e5468f599334a0aa) I have my protocols, the file `intercom.py` contains also the protocol definition.

### Data protocol
A `Publisher` class to publish data, a `Listener` class which listens on published data (and control commands sent via the command protocol) and emits a Qt Signal.

In the `proxy_server.py` file is the proxy server, to which publisher and subsciber (Listener) connect. There is also a proxy server which connects two local proxy servers, allowing to stream data from one computer to another one. Further, there are convenience functions to start these severs in another thread and to start one with command line arguments executing the file (creating an operating system starter).

With this data protocol, it is easy to have a "Republisher", which listens to data, transforms it and sends it modified (for example takes a photo diode voltage and publishes the corresponding light power according to some calibration function). That way, you do not have to modify either the instrument nor the data acquisition.

### Control protocol
It uses the same Listener which binds to a fixed port. A `Communicator` connects to that port and sends pickled python object. The interpretation of these objects is not defined in the protocol.

Example: My data acquisition program expects a dictionary and sets the properties according to the dictionary.
